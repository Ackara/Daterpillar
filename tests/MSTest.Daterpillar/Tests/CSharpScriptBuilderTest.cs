using Ackara.Daterpillar.Scripting;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shouldly;
using System;
using System.Collections.Generic;
using System.Linq;

namespace MSTest.Daterpillar.Tests
{
    [TestClass]
    [UseReporter(typeof(FileLauncherReporter), typeof(ClipboardReporter))]
    public class CSharpScriptBuilderTest
    {
        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_class_when_all_settings_are_enabled()
        {
            // Arrange
            var schema = MockData.GetSchema(FName.scriptingTest_partial_schemaXML);
            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                AddConstants = true,
                IgnoreComments = false,
                AddSchemaAttributes = true,
                UseVirtualProperties = true,
                InheritEntityBaseClass = true,
                AddDataContractAttributes = true,
            });

            // Act
            sut.Append(schema);
            var script = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref script, out string errorMsg);

            // Assert
            script.ShouldNotBeNullOrEmpty();
            Approvals.Verify(script);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_class_when_all_settings_are_disabled()
        {
            // Arrange
            var schema = MockData.GetSchema(FName.scriptingTest_partial_schemaXML);
            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                AddConstants = false,
                IgnoreComments = true,
                AddSchemaAttributes = false,
                UseVirtualProperties = false,
                InheritEntityBaseClass = false,
                AddDataContractAttributes = false,
            });

            // Act
            sut.Append(schema);
            var script = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref script, out string errorMsg);

            // Assert
            script.ShouldNotBeNullOrEmpty();
            Approvals.Verify(script);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_enum()
        {
            // Arrange
            var sample = new Dictionary<string, int>()
            {
                {"Red", 0 },
                {"Green", 1 },
                {"Blue", 2 },
                {"Dark Gray", 4 },
            };

            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                IgnoreComments = false,
                AddSchemaAttributes = true,
                UseVirtualProperties = true,
                InheritEntityBaseClass = true,
            });

            // Act
            sut.Append("Color", sample);
            var code = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref code, out string errorMsg);

            // Assert
            code.ShouldNotBeNullOrEmpty();
            Approvals.Verify(code);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        [TestMethod]
        public void Append_should_return_a_csharp_script_that_defines_a_enum_that_supports_flags()
        {
            // Arrange
            var sample = new Dictionary<string, int>()
            {
                {"Red", 0 },
                {"Green", 1 },
                {"Blue", 2 },
                {"Dark Gray", 4 },
            };

            var sut = new CSharpScriptBuilder(new CSharpScriptBuilderSettings()
            {
                IgnoreComments = false,
                AddSchemaAttributes = true,
                UseVirtualProperties = true,
                InheritEntityBaseClass = true,
            });

            // Act
            sut.Append("Color", sample, flag: true);
            var code = sut.GetContent();
            var codeIsCompilable = TryCompilation(ref code, out string errorMsg);

            // Assert
            code.ShouldNotBeNullOrEmpty();
            Approvals.Verify(code);
            codeIsCompilable.ShouldBeTrue(errorMsg);
        }

        private static bool TryCompilation(ref string code, out string errorMsg)
        {
            var errors = CSharpSyntaxTree.ParseText(code).GetDiagnostics();
            errorMsg = string.Join(Environment.NewLine, errors.Select(x => $"[{x.Severity}] {x.GetMessage()}"));

            if (errors.Count() > 0)
            {
                code = string.Format("/*{0}ERRORS!!!{0}{1}{0}*/{0}{0}{2}", Environment.NewLine, errorMsg, code);
            }

            return errors.Count() == 0;
        }
    }
}