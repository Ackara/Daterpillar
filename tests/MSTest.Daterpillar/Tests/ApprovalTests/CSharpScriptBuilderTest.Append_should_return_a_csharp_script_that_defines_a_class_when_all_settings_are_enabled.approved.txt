/// <summary>
/// Represents a card_type.
/// </summary>
[Table("card_type")]
public partial class CardType
{
	#region SCHEMA NAMES

	public const string Table = "card_type";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "int")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 32)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

}

/// <summary>
/// Represents a attribute.
/// </summary>
[Table("attribute")]
public partial class Attribute
{
	#region SCHEMA NAMES

	public const string Table = "attribute";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "tinyInt")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 16)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

}

/// <summary>
/// Represents a card_icon.
/// </summary>
[Table("card_icon")]
public partial class CardIcon
{
	#region SCHEMA NAMES

	public const string Table = "card_icon";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "tinyInt")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 16)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

}

/// <summary>
/// Represents a monster_type.
/// </summary>
[Table("monster_type")]
public partial class MonsterType
{
	#region SCHEMA NAMES

	public const string Table = "monster_type";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "tinyInt")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 32)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

}

/// <summary>
/// Represents a monster's ability.
/// </summary>
[Table("ability")]
public partial class Ability
{
	#region SCHEMA NAMES

	public const string Table = "ability";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Get or set the Id. This value is a flag/it must be a power of 2.
	/// </summary>
	[Column("Id", "int")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 32)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

}

/// <summary>
/// Represents a card.
/// </summary>
[Table("card")]
public partial class Card
{
	#region SCHEMA NAMES

	public const string Table = "card";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	public const string TextColumn = "Text";

	public const string PendulumEffectColumn = "Pendulum_Effect";

	public const string LevelColumn = "Level";

	public const string AtkColumn = "Atk";

	public const string DefColumn = "Def";

	public const string CardTypeIdColumn = "Card_Type_Id";

	public const string AttributeIdColumn = "Attribute_Id";

	public const string CardIconIdColumn = "Card_Icon_Id";

	public const string MonsterTypeIdColumn = "Monster_Type_Id";

	public const string AbilityIdColumn = "Ability_Id";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "int")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 128)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

	/// <summary>
	/// Gets or Sets the Text.
	/// </summary>
	[Column("Text", "text")]
	public virtual string Text { get; set; }

	/// <summary>
	/// Gets or Sets the Pendulum_Effect.
	/// </summary>
	[Column("Pendulum_Effect", "text")]
	public virtual string PendulumEffect { get; set; }

	/// <summary>
	/// Gets or Sets the Level.
	/// </summary>
	[Column("Level", "tinyInt", DefaultValue = "0")]
	public virtual int Level { get; set; }

	/// <summary>
	/// Gets or Sets the Atk.
	/// </summary>
	[Column("Atk", "int")]
	public virtual int Atk { get; set; }

	/// <summary>
	/// Gets or Sets the Def.
	/// </summary>
	[Column("Def", "int")]
	public virtual int Def { get; set; }

	/// <summary>
	/// Gets or Sets the Card_Type_Id.
	/// </summary>
	[Column("Card_Type_Id", "int")]
	public virtual int CardTypeId { get; set; }

	/// <summary>
	/// Gets or Sets the Attribute_Id.
	/// </summary>
	[Column("Attribute_Id", "tinyInt")]
	[Index(IndexType.Index)]
	[ForeignKey("attribute", "Id", OnUpdate = ReferentialAction.Cascade, OnDelete = ReferentialAction.Restrict)]
	public virtual int AttributeId { get; set; }

	/// <summary>
	/// Gets or Sets the Card_Icon_Id.
	/// </summary>
	[Column("Card_Icon_Id", "tinyInt")]
	[Index(IndexType.Index)]
	[ForeignKey("card_icon", "Id", OnUpdate = ReferentialAction.Cascade)]
	public virtual int CardIconId { get; set; }

	/// <summary>
	/// Gets or Sets the Monster_Type_Id.
	/// </summary>
	[Column("Monster_Type_Id", "tinyInt")]
	[Index(IndexType.Index)]
	[ForeignKey("monster_type", "Id", OnUpdate = ReferentialAction.Cascade, OnDelete = ReferentialAction.Restrict)]
	public virtual int MonsterTypeId { get; set; }

	/// <summary>
	/// Gets or Sets the Ability_Id.
	/// </summary>
	[Column("Ability_Id", "int")]
	public virtual int AbilityId { get; set; }

}

/// <summary>
/// Represents a card_extras.
/// </summary>
[Table("card_extras")]
public partial class CardExtras
{
	#region SCHEMA NAMES

	public const string Table = "card_extras";

	public const string CardIdColumn = "Card_Id";

	public const string RulingsColumn = "Rulings";

	public const string TipsColumn = "Tips";

	public const string TriviaColumn = "Trivia";

	public const string PasscodeColumn = "Passcode";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Card_Id.
	/// </summary>
	[Column("Card_Id", "int")]
	[Index(IndexType.PrimaryKey)]
	[ForeignKey("card", "Id")]
	public virtual int CardId { get; set; }

	/// <summary>
	/// Gets or Sets the Rulings.
	/// </summary>
	[Column("Rulings", "text", Nullable = true)]
	public virtual string Rulings { get; set; }

	/// <summary>
	/// Gets or Sets the Tips.
	/// </summary>
	[Column("Tips", "text", Nullable = true)]
	public virtual string Tips { get; set; }

	/// <summary>
	/// Gets or Sets the Trivia.
	/// </summary>
	[Column("Trivia", "text", Nullable = true)]
	public virtual string Trivia { get; set; }

	/// <summary>
	/// Gets or Sets the Passcode.
	/// </summary>
	[Column("Passcode", "int")]
	public virtual int Passcode { get; set; }

}

/// <summary>
/// Represents a pack.
/// </summary>
[Table("pack")]
public partial class Pack
{
	#region SCHEMA NAMES

	public const string Table = "pack";

	public const string IdColumn = "Id";

	public const string AcronymColumn = "Acronym";

	public const string NameColumn = "Name";

	public const string DescriptionColumn = "Description";

	public const string ReleaseDateColumn = "Release_Date";

	public const string SizeColumn = "Size";

	public const string KonamiIdColumn = "Konami_Id";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "int")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Acronym.
	/// </summary>
	[Column("Acronym", "varchar", 32)]
	public virtual string Acronym { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 128)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

	/// <summary>
	/// Gets or Sets the Description.
	/// </summary>
	[Column("Description", "text")]
	public virtual string Description { get; set; }

	/// <summary>
	/// Gets or Sets the Release_Date.
	/// </summary>
	[Column("Release_Date", "date")]
	public virtual DateTime ReleaseDate { get; set; }

	/// <summary>
	/// The number of unique cards in the pack.
	/// </summary>
	[Column("Size", "int")]
	public virtual int Size { get; set; }

	/// <summary>
	/// Gets or Sets the Konami_Id.
	/// </summary>
	[Column("Konami_Id", "int")]
	public virtual int KonamiId { get; set; }

}

/// <summary>
/// Represents a rarity.
/// </summary>
[Table("rarity")]
public partial class Rarity
{
	#region SCHEMA NAMES

	public const string Table = "rarity";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	public const string CodeColumn = "Code";

	public const string ValueColumn = "Value";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Id.
	/// </summary>
	[Column("Id", "tinyInt")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 32)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

	/// <summary>
	/// Gets or Sets the Code.
	/// </summary>
	[Column("Code", "varchar", 4)]
	public virtual string Code { get; set; }

	/// <summary>
	/// Gets or Sets the Value.
	/// </summary>
	[Column("Value", "int")]
	public virtual int Value { get; set; }

}

/// <summary>
/// Represents a card_number.
/// </summary>
[Table("card_number")]
public partial class CardNumber
{
	#region SCHEMA NAMES

	public const string Table = "card_number";

	public const string PackAcronymColumn = "Pack_Acronym";

	public const string NumberColumn = "Number";

	public const string RarityIdColumn = "Rarity_Id";

	public const string ArtIndexColumn = "Art_Index";

	public const string CardIdColumn = "Card_Id";

	public const string PackIdColumn = "Pack_Id";

	#endregion SCHEMA NAMES

	/// <summary>
	/// Gets or Sets the Pack_Acronym.
	/// </summary>
	[Column("Pack_Acronym", "varchar", 32)]
	[Index(IndexType.PrimaryKey)]
	public virtual string PackAcronym { get; set; }

	/// <summary>
	/// Gets or Sets the Number.
	/// </summary>
	[Column("Number", "smallInt")]
	[Index(IndexType.PrimaryKey)]
	public virtual int Number { get; set; }

	/// <summary>
	/// Gets or Sets the Rarity_Id.
	/// </summary>
	[Column("Rarity_Id", "tinyInt")]
	[Index(IndexType.Index)]
	[Index(IndexType.PrimaryKey)]
	[ForeignKey("rarity", "Id")]
	public virtual int RarityId { get; set; }

	/// <summary>
	/// Gets or Sets the Art_Index.
	/// </summary>
	[Column("Art_Index", "tinyInt")]
	[Index(IndexType.PrimaryKey)]
	public virtual int ArtIndex { get; set; }

	/// <summary>
	/// Gets or Sets the Card_Id.
	/// </summary>
	[Column("Card_Id", "int")]
	[Index(IndexType.Index)]
	[ForeignKey("card", "Id")]
	public virtual int CardId { get; set; }

	/// <summary>
	/// Gets or Sets the Pack_Id.
	/// </summary>
	[Column("Pack_Id", "int")]
	[Index(IndexType.Index)]
	[ForeignKey("pack", "Id")]
	public virtual int PackId { get; set; }

}

/// <summary>
/// A custom table comment.
/// </summary>
[Table("effect")]
public partial class Effect
{
	#region SCHEMA NAMES

	public const string Table = "effect";

	public const string IdColumn = "Id";

	public const string NameColumn = "Name";

	#endregion SCHEMA NAMES

	/// <summary>
	/// A custom column comment.
	/// </summary>
	[Column("Id", "int", AutoIncrement = true)]
	public virtual int Id { get; set; }

	/// <summary>
	/// Gets or Sets the Name.
	/// </summary>
	[Column("Name", "varchar", 32)]
	[Index(IndexType.Index, Unique = true)]
	public virtual string Name { get; set; }

}

